### PR TITLE
Do not skip tests silently

### PR DESCRIPTION
**Pull request by olivergondza**
Replace early returns in tests with assumptions.

There should be 15 new skipped tests after this refactorization.

### CODE REVIEW

File: core/src/test/java/hudson/FilePathTest.java [hunk 1]
```diff
@@ -464,7 +464,7 @@ private void checkTarUntarRoundTrip(String filePrefix, long fileSize) throws Exc
     }
 
     @Test public void symlinkInTar() throws Exception {
-        if (Functions.isWindows())  return; // can't test on Windows
+        assumeTrue(!Functions.isWindows()); // can't test on Windows
```
Comment (KostyaSha) [inline core/src/test/java/hudson/FilePathTest.java]: then test will fail? Should it be ignorefailed?

Comment (olivergondza) [inline core/src/test/java/hudson/FilePathTest.java]: Then the test will be *skipped* in JUnit (and Jenkins), while before it was considered successful even though it in fact did not run.

File: test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java [hunk 2]
```diff
@@ -92,10 +93,9 @@ public void testFreeStyleShellOnSlave() throws Exception {
 	}
 
 	public void testFreeStyleAntOnSlave() throws Exception {
-        if (jenkins.getDescriptorByType(Ant.DescriptorImpl.class).getInstallations().length == 0) {
-            System.out.println("Cannot do testFreeStyleAntOnSlave without ANT_HOME");
-            return;
-        }
+	    // Cannot do testFreeStyleAntOnSlave without ANT_HOME
+	    Assume.assumeTrue((jenkins.getDescriptorByType(Ant.DescriptorImpl.class).getInstallations().length > 0);
+
```
Comment (KostyaSha) [inline test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java]: intend not consistent?

File: core/src/test/java/hudson/FilePathTest.java (hunk 1 above)
Comment (KostyaSha) [inline core/src/test/java/hudson/FilePathTest.java]: Why not `assumeFalse("can't test on Windows", Functions.isWindows())` ?

**Comment by KostyaSha** [comment]
:bug: for not using `assumeFalse(String message, boolean b)`

File: core/src/test/java/hudson/util/ProcessTreeTest.java [hunk 3]
```diff
@@ -27,8 +29,7 @@
     
     @Test public void remoting() throws Exception {
         // on some platforms where we fail to list any processes, this test will just not work
-        if (ProcessTree.get()==ProcessTree.DEFAULT)
-            return;
+        Assume.assumeTrue(ProcessTree.get()!=ProcessTree.DEFAULT);
```
Comment (jtnord) [inline core/src/test/java/hudson/util/ProcessTreeTest.java]: I find `assumeThat(ProcessTree.get(), not(sameInstance(ProcessTree.DEFAULT)))` is more readable that a true followed by a negative

File: test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java [hunk 4]
```diff
@@ -92,10 +93,9 @@ public void testFreeStyleShellOnSlave() throws Exception {
 	}
 
 	public void testFreeStyleAntOnSlave() throws Exception {
-        if (jenkins.getDescriptorByType(Ant.DescriptorImpl.class).getInstallations().length == 0) {
-            System.out.println("Cannot do testFreeStyleAntOnSlave without ANT_HOME");
-            return;
-        }
+	    // Cannot do testFreeStyleAntOnSlave without ANT_HOME
+	    Assume.assumeTrue((jenkins.getDescriptorByType(Ant.DescriptorImpl.class).getInstallations().length > 0);
```
Comment (jtnord) [inline test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java]: `assumeThat(jenkins.getDescriptorByType(Ant.DescriptorImpl.class).getInstallations(), not(empty()))`

File: core/src/test/java/hudson/util/ProcessTreeTest.java (hunk 3 above)
Comment (KostyaSha) [inline core/src/test/java/hudson/util/ProcessTreeTest.java]: @jtnord :+1: `assumeThat(String message, T actual, Matcher<T> matcher) `

File: core/src/test/java/hudson/FilePathTest.java (hunk 1 above)
Comment (jtnord) [inline core/src/test/java/hudson/FilePathTest.java]: there is no `assumeFalse` in junit.

Comment (KostyaSha) [inline core/src/test/java/hudson/FilePathTest.java]: http://junit.org/apidocs/org/junit/Assume.html#assumeFalse(java.lang.String, boolean)

Comment (KostyaSha) [inline core/src/test/java/hudson/FilePathTest.java]: https://github.com/jenkinsci/jenkins/blob/master/pom.xml#L151-L155 in 4.11 exists
Right after `assumeTrue` and before `assumeThat`

**Comment by jtnord** [comment]
:+1: on the change - but there are a few nits - but better to have skipped tests than tests that look like they pass.

When testing a larger expression it is often more readable (and anefficient) to use [hamcrest](http://hamcrest.org/JavaHamcrest/javadoc/1.3/) matchers.

I also find that it hurts my head when I see `assumeTrue(!xyz)` due to the lack of and `assumeFalse()`              
could be rewritten as `assumeThat(Functions.isWindows(), is(false))` or even `assumeThat("test will not run on windows platforms", Functions.isWindows(), is(false))`

**Comment by KostyaSha** [comment]
> I also find that it hurts my head when I see assumeTrue(!xyz) due to the lack of and assumeFalse()

@jtnord https://github.com/junit-team/junit/blob/r4.11/src/main/java/org/junit/Assume.java#L61-L68

**Comment by jtnord** [comment]
@KostyaSha - even better.  I was trusting google and it was giving me [this page](http://junit.sourceforge.net/javadoc/org/junit/Assume.html)

**Comment by KostyaSha** [comment]
> this page

http://junit.sourceforge.net/  `Please see our main site at junit.org`

**Comment by KostyaSha** [comment]
And from Assume javadoc:
```
These methods can be used directly: Assume.assumeTrue(...), 
however, they read better if they are referenced through static import:
 import static org.junit.Assume.*;
    ...
    assumeTrue(...);
```

**Comment by olivergondza** [comment]
Ad using `assertFalse`: right, my IDE mislead me those methods are not there in version we are using.

As static/star imports: we do not have any clear policy here.

File: test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java [hunk 5]
```diff
@@ -92,10 +93,9 @@ public void testFreeStyleShellOnSlave() throws Exception {
 	}
 
 	public void testFreeStyleAntOnSlave() throws Exception {
-        if (jenkins.getDescriptorByType(Ant.DescriptorImpl.class).getInstallations().length == 0) {
-            System.out.println("Cannot do testFreeStyleAntOnSlave without ANT_HOME");
-            return;
-        }
+		// Cannot do testFreeStyleAntOnSlave without ANT_HOME
+	    Assume.assumeFalse(jenkins.getDescriptorByType(Ant.DescriptorImpl.class).getInstallations().length == 0);
```
Comment (jtnord) [inline test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java]: need to indent with tabs here :-)

**Comment by KostyaSha** [comment]
> As static/star imports: we do not have any clear policy here.

Why do you need policy? I see static import for Assert and non-static for Assume. I trust javadoc about readability.

CC @lanwen

File: test/src/test/java/hudson/tools/JDKInstallerTest.java [hunk 6]
```diff
@@ -149,8 +150,7 @@ public void autoInstallation142_17() throws Exception {
      */
     private void doTestAutoInstallation(String id, String fullversion) throws Exception {
         // this is a really time consuming test, so only run it when we really want
-        if(!Boolean.getBoolean("jenkins.testJDKInstaller"))
-            return;
+        Assume.assumeTrue(Boolean.getBoolean("jenkins.testJDKInstaller"));
```
Comment (KostyaSha) [inline test/src/test/java/hudson/tools/JDKInstallerTest.java]: Place string description in assume

File: test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java (hunk 5 above)
Comment (KostyaSha) [inline test/src/test/java/hudson/tasks/EnvVarsInConfigTasksTest.java]: and migrate `Cannot do testFreeStyleAntOnSlave without ANT_HOME` into assume argument

**Comment by lanwen** [comment]
I'm completely agree with @jtnord for using `assumeThat("description", actual, matcherFromHamcrest)` as more detailed, readable and refactorible.

For example what do you think about assumption if you have only 
> symlinkInTar, skipped. 
Expect true, was false
in remote CI build? Think most of users have no idea what that means.

With matchers and description it can be 
> symlinkInTar, skipped.
"This test is not for windows", 
expect: list has size greater than 0,
but was: 0

about static imports in tests - when you write tests you should write case as more human readable, as possible. Static imports can help you to give any possible name to any class (as factory) or for static function (to hide it test-code mentions about enclosing class). You should write some kind of dsl with help of java. In this case anyone can understand the case and change it if it fail.

**Comment by lanwen** [comment]
As of you use this assume too often (and keep your code DRY), it can be done with help of `org.jvnet.hudson.test.JenkinsRecipe`

```java
@Documented
@Recipe(WithTimeout.RunnerImpl.class)
@JenkinsRecipe(NotForWindows.RuleRunnerImpl.class)
@Target(METHOD)
@Retention(RUNTIME)
public @interface NotForWindows {
    class RuleRunnerImpl extends JenkinsRecipe.Runner<NotForWindows> {
        @Override
        public void setup(JenkinsRule jenkinsRule, NotForWindows recipe) throws Exception {
            assumeThat("Whis test can not be run on windows", isWindows(), is(false));
        }
    }
}
```

After that, test can be annotated with `@NotForWindows` to do same effect.

**Comment by olivergondza** [comment]
I do not really understand where this discussion is going. I do not suggest anybody to use my *preferred* import style so we should either agree on one (not in this PR please) and use it everywhere in Jenkins core or accept the fact the our code is a mixture of different styles (which it is). FTR, we are not even consistent in using static imports for `org.junit.Assert`.

Ad hamcrest, yes it is superior to `org.junit.Assume`, though I reject replacing `assertFalse(X)` with `assertThat(X, is(false))` blindly as it has no practice advantage. Also, hamcrest can be lot more useful for asserting (then assume*), I have not seen anybody to push on using it so strongly before...

Ad `NotForWindows`: Good point.

**Comment by olivergondza** [comment]
> Ad  NotForWindows : Good point.

Is there any way to share this between core tests and jenkins-test-harness? Does not seem to fit in lib-test-annotations.

**Comment by lanwen** [comment]
`assumeThat(X, is(false))` can be easily changed to `assumeThat(X, not(false))` or to `assumeThat(X, both(withY(is(false))).and(withZ(is(true)))` if condition became to another form. It is more readable in review when changed only matcher and not whole assertion.

File: test/src/test/groovy/jenkins/model/PeepholePermalinkTest.groovy [hunk 7]
```diff
@@ -71,7 +73,7 @@ class PeepholePermalinkTest {
      */
     @Test
     void legacyCompatibility() {
-        if (Functions.isWindows())  return; // can't run on windows because we rely on symlinks
+        Assume.assumeFalse("can't run on windows because we rely on symlinks", Functions.isWindows());
```
Comment (jtnord) [inline test/src/test/groovy/jenkins/model/PeepholePermalinkTest.groovy]: Hmm. this is actually wrong - windows does have symlink support in jenkins.
[if it is enabled](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/Util.java#L1265) for the current user is a different issue.
This is not a `-1` on this as this is like for like replacement.

**Comment by olivergondza** [comment]
As `NotForWindows` seems out of scope of this PR, I am removing WIP label.

**Comment by jtnord** [comment]
:+1:

**Comment by KostyaSha** [comment]
> though I reject replacing assumeFalse(X) with assumeThat(X, is(false)) blindly as it has no practice advantage.

AssumeFalse is not readable for objects and @jtnord first of all pointed to use assumeThat for objects and mistakenly/optionally suggested for replacing it for booleans.

File: core/src/test/java/hudson/util/ProcessTreeTest.java [hunk 8]
```diff
@@ -26,9 +28,7 @@
     }
     
     @Test public void remoting() throws Exception {
-        // on some platforms where we fail to list any processes, this test will just not work
-        if (ProcessTree.get()==ProcessTree.DEFAULT)
-            return;
+        Assume.assumeFalse("on some platforms where we fail to list any processes", ProcessTree.get()==ProcessTree.DEFAULT);
```
Comment (KostyaSha) [inline core/src/test/java/hudson/util/ProcessTreeTest.java]: `assumeThat()`

File: test/src/test/java/hudson/slaves/JNLPLauncherTest.java [hunk 9]
```diff
@@ -31,36 +31,43 @@
...
     /**
      * Starts a JNLP slave agent and makes sure it successfully connects to Hudson. 
      */
+    @Test
     public void testLaunch() throws Exception {
-        if(GraphicsEnvironment.isHeadless()) {
-            System.err.println("Skipping JNLPLauncherTest.testLaunch because we are running headless");
-            return;
-        }
+        Assume.assumeFalse("Skipping JNLPLauncherTest.testLaunch because we are running headless", GraphicsEnvironment.isHeadless());
```
Comment (KostyaSha) [inline test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: I think test case will write that test is skipped and here required only description that can be short
i.e. `JNLPLauncherTest.testLaunch was in headless mode`

Comment (jtnord) [inline test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: no point in putting the test name as it is available in the test report ([example](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/lastCompletedBuild/testReport/hudson/FunctionsTest/testGetRelativeLinkTo_JobNotContainedInView/)) and would be incorrect in any refactoring.

**Comment by jtnord** [comment]
@KostyaSha whilst this change is not perfect - it is (IMO) far better than the code that existed before - can we agree this is better than before and get this included?

:+1:

**Comment by KostyaSha** [comment]
@jtnord of course it better but i don't understand why it is a problem to make it more better.

File: test/src/test/java/hudson/slaves/JNLPLauncherTest.java (hunk 9 above)
Comment (KostyaSha) [inline test/src/test/java/hudson/slaves/JNLPLauncherTest.java]: Right, then probably `GraphicsEnvironment was in headless mode`?

**Pull request merged**